### PR TITLE
(#232,#235,#236,#240) .NET SDK v3.0.2 items

### PR DESCRIPTION
**Pull request by adrianhall**

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@adrianhall__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (adrianha). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: README.md [hunk 1]
```diff
@@ -13,7 +13,18 @@ To get the source code of our SDKs and samples via **git** just type:
     git clone https://github.com/Azure/azure-mobile-apps-net-client.git
     cd ./azure-mobile-apps-net-client/
 
-## Reference Documentation
+## Supported platforms
+
+* Xamarin Android releases for API 19 through 24 (KitKat through Nougat)
```
Comment (adrianhall) [inline README.md]: I don't think we can say API 19+, as API 24 proved.  I'd rather do the work on each new version and have to do a quick README update.

File: README.md [hunk 2]
```diff
@@ -13,7 +13,18 @@ To get the source code of our SDKs and samples via **git** just type:
...
+* Xamarin iOS releases for iOS versions 8.0 through 10.0
+* Universal Windows Platform
+* Windows Phone 8.1
+* Windows Phone 8.0 except for Silverlight applications.
```
Comment (adrianhall) [inline README.md]: Duh - yeah - will correct.  Thanks.

File: README.md [hunk 3]
```diff
@@ -13,7 +13,18 @@ To get the source code of our SDKs and samples via **git** just type:
     git clone https://github.com/Azure/azure-mobile-apps-net-client.git
     cd ./azure-mobile-apps-net-client/
 
-## Reference Documentation
+## Supported platforms
+
+* Xamarin Android releases for API 19 through 24 (KitKat through Nougat)
+* Xamarin iOS releases for iOS versions 8.0 through 10.0
```
Comment (adrianhall) [inline README.md]: As with the Android version, unless Apple is committing to testing our library, we need to ensure compatibility before we support it.

**Comment by adrianhall** [comment]
Re: the PCLStorage.  If I do the System.IO, I have to add a Platform specific IO package in.  I was going along the same lines.  However, SQLiteStore is a PCL, so it doesn't get the entire System.IO - most notably, File.* are not there, so I can't do File.Create or File.Exists.  

The way around this is to add another method (EnsureDatabaseExists(fullpathname)) to the IPlatform.cs and do it there.  That would work around the issue.  I'll discuss internally on Monday and see which direction we want to go.  (@dhei - if you have comment over the weekend, please add it)

**Comment by jamesmontemagno** [comment]
I would most likely just want to validate support from @dplaisted on it. It think that going against System.IO and controlling it is worth it as taking a dependency on PLCStorage brings in even more DLLs, I think it would be better to bring in platform specific DLLs to access System.IO and actually for 4.0 they will go away with .NET Standard

File: README.md (hunk 2 above)
Comment (dhei) [inline README.md]: yeah, it's Windows Phone Silverlight 8.1

**Comment by dhei** [comment]
@adrianhall: I agree with @jamesmontemagno. I think it's easier to manage platform specific IO package than adding a new PCL package. More importantly, [PCLStorage](https://www.nuget.org/packages/pclstorage) takes dependency on Microsoft.Bcl >= 1.1.6 but we're planning to completely get rid of Microsoft.Bcl when we moves to .NET Standard 1.3.

File: src/Microsoft.WindowsAzure.MobileServices.SQLiteStore/MobileServiceSQLiteStore.cs [hunk 4]
```diff
@@ -44,8 +45,27 @@ public MobileServiceSQLiteStore(string fileName)
...
+
+                // Ensure the file exists (synchronous deliberately since we are in constructor)
+                var folderRef = FileSystem.Current.GetFolderFromPathAsync(directoryName).Result;
+                try
+                {
+                    folderRef.CreateFileAsync(nFileName, CreationCollisionOption.FailIfExists).Wait();
+                }
+                catch (FileNotFoundException)
+                {
+                    // Ignore this error - it means the file exists
```
Comment (adrianhall) [inline src/Microsoft.WindowsAzure.MobileServices.SQLiteStore/MobileServiceSQLiteStore.cs]: Ok - irrelevant now since I've factored it out.

**Comment by adrianhall** [comment]
Per the review comments, I've added a per-platform method called "EnsureFileExists(path)".  On Windows, it uses Windows.Storage (since System.IO.File does not exist - go figure), and on iOS/Android, it uses System.IO.File.  I've also removed the dependency on PCLStorage from the SQLiteStore. 

This does increase our public API footprint on the MobileServiceClient, but I think both pieces are likely to be useful for anyone implementing a local store provider, so it's a good addition.  

Comments appreciated.

File: Microsoft.WindowsAzure.Mobile.Managed_Full_IncludeXamarin.sln [hunk 5]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2013
-VisualStudioVersion = 12.0.40629.0
```
Comment (dhei) [inline Microsoft.WindowsAzure.Mobile.Managed_Full_IncludeXamarin.sln]: Microsoft.WindowsAzure.Mobile.Managed_Full_IncludeXamarin.sln is supposed work with Visual Studio 2013 and Microsoft.WindowsAzure.Mobile.Managed_Full_IncludeXamarin_V15.sln is for Visual Studio 2015. So can you make this change in V15 sln file?

Comment (dhei) [inline Microsoft.WindowsAzure.Mobile.Managed_Full_IncludeXamarin.sln]: The rest looks good to me.

**Comment by dhei** [comment]
>     public class CurrentPlatform : IPlatform

Please add EnsureFileExists() to this class.

Error CS0535 'CurrentPlatform' does not implement interface member 'IPlatform.EnsureFileExists(string)' Microsoft.WindowsAzure.Mobile.Ext.Android 


---
Refers to: src/Microsoft.WindowsAzure.MobileServices.Android/Platform/CurrentPlatform.cs:12 in b90a6d0. [](commit_id = b90a6d07b3e63d78674b0ab2b0280f8a463c9b8d, deletion_comment = False)

**Pull request merged**